compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -547,159 +519,156 @@ private boolean isPublicInvokableNode(BLangInvokableNode invNode) {
     }
 
     @Override
-    public void visit(BLangBlockFunctionBody body) {
-        boolean prevWithinTxScope = withinTransactionScope;
-        boolean prevLoopAlterNotAllowed = loopAlterNotAllowed;
-        loopAlterNotAllowed = loopCount > 0;
-        if (!transactionalFuncCheckStack.empty() && !withinTransactionScope) {
-            withinTransactionScope = transactionalFuncCheckStack.peek();
-        }
-        final SymbolEnv blockEnv = SymbolEnv.createFuncBodyEnv(body, env);
+    public void visit(BLangBlockFunctionBody body, AnalyzerData data) {
+        boolean prevWithinTxScope = data.withinTransactionScope;
+        boolean prevLoopAlterNotAllowed = data.loopAlterNotAllowed;
+        data.loopAlterNotAllowed = data.loopCount > 0;
+        if (!prevWithinTxScope) {
+            data.withinTransactionScope = data.transactionalFuncCheckStack.peek();
+        }
+        data.env = SymbolEnv.createFuncBodyEnv(body, data.env);
         for (BLangStatement e : body.stmts) {
-            this.inInternallyDefinedBlockStmt = true;
-            analyzeNode(e, blockEnv);
+            data.inInternallyDefinedBlockStmt = true;
+            analyzeNode(e, data);
         }
-        this.inInternallyDefinedBlockStmt = false;
-        if (!transactionalFuncCheckStack.empty() && transactionalFuncCheckStack.peek()) {
-            withinTransactionScope = prevWithinTxScope;
+        data.inInternallyDefinedBlockStmt = false;
+        if (data.transactionalFuncCheckStack.peek()) {
+            data.withinTransactionScope = prevWithinTxScope;
         }
-        loopAlterNotAllowed = prevLoopAlterNotAllowed;
+        data.loopAlterNotAllowed = prevLoopAlterNotAllowed;
     }
 
     @Override
-    public void visit(BLangExprFunctionBody body) {
-        analyzeExpr(body.expr);
+    public void visit(BLangExprFunctionBody body, AnalyzerData data) {
+        analyzeExpr(body.expr, data);
     }
 
     @Override
-    public void visit(BLangExternalFunctionBody body) {
+    public void visit(BLangExternalFunctionBody body, AnalyzerData data) {
         // do nothing
     }
 
     @Override
-    public void visit(BLangForkJoin forkJoin) {
+    public void visit(BLangForkJoin forkJoin, AnalyzerData data) {
         if (forkJoin.workers.isEmpty()) {
             dlog.error(forkJoin.pos, DiagnosticErrorCode.INVALID_FOR_JOIN_SYNTAX_EMPTY_FORK);
         }
     }
 
     @Override
-    public void visit(BLangTransaction transactionNode) {
+    public void visit(BLangTransaction transactionNode, AnalyzerData data) {
         //Check whether transaction statement occurred in a transactional scope
-        if (!transactionalFuncCheckStack.empty() && transactionalFuncCheckStack.peek()) {
+        if (data.transactionalFuncCheckStack.peek()) {
```

Comment:
Don't we need the empty check anymore?

Reply:
https://github.com/ballerina-platform/ballerina-lang/pull/34909#discussion_r821554740